runtime/src/mem/jemalloc/mem-jemalloc.c

```diff
@@ -30,44 +30,198 @@
+  // NULL on error
+  // """
+  if (chunk && chunk != p) {
+    return NULL;
+  }
+
+  p_offset = (uintptr_t)p - (uintptr_t)(heap_base);
+
+  size = ((size + alignment - 1) / alignment) * alignment;
+
+  // If we're out of space on the heap, return NULL
+  if (size > heap_size - p_offset) {
+    return NULL;
+  }
+
+  // Update the current pointer, now that we've past any early returns.
+  cur_heap_offset = p_offset + size;
+
+  // Zero if we were told to
+  if (*zero) {
+     memset(p, 0, size);
+  }
+
+  // Commit is not relevant for linux/darwin, but jemalloc wants us to set it
+  *commit = true;
+
+  return p;
+}
+
+//
+// Returning true indicates an opt-out of these hooks. For dalloc, this means
+// that we never get memory back from jemalloc and we just let it re-use it in
+// the future.
+//
+static bool null_dalloc(void *chunk, size_t size, bool committed, unsigned arena_ind) {
+  return true;
+}
+static bool null_commit(void *chunk, size_t size, size_t offset, size_t length, unsigned arena_ind) {
+  return true;
+}
+static bool null_decommit(void *chunk, size_t size, size_t offset, size_t length, unsigned arena_ind) {
+  return true;
+}
+static bool null_purge(void *chunk, size_t size, size_t offset, size_t length, unsigned arena_ind) {
+  return true;
+}
+static bool null_split(void *chunk, size_t size, size_t size_a, size_t size_b, bool committed, unsigned arena_ind) {
+  return true;
+}
+static bool null_merge(void *chunk_a, size_t size_a, void *chunk_b, size_t size_b, bool committed, unsigned arena_ind) {
+  return true;
+}
+// *** End chunk hook replacements *** //
+
+
+// TODO check for errors on all je_mallctl calls
+
+// get the number of arenas could use
+static unsigned get_num_arenas(void) {
+  unsigned narenas;
+  size_t sz;
+  sz = sizeof(narenas);
+  je_mallctl("opt.narenas", &narenas, &sz, NULL, 0);
+  return narenas;
+}
+
+// initialize our arenas (this is required to be able to set the chunk hooks)
+static void initialize_arenas(void) {
+  unsigned narenas;
+  unsigned arena;
+
+  // for each arena, set the current thread to use it (this initializes each arena)
+  narenas = get_num_arenas();
+  for (arena=0; arena<narenas; arena++) {
+    je_mallctl("thread.arena", NULL, NULL, &arena, sizeof(arena));
+  }
+
+  // then set the current thread back to using arena 0
+  arena = 0;
+  je_mallctl("thread.arena", NULL, NULL, &arena, sizeof(arena));
+}
+
+
+// replace the chunk hooks for each arena with the hooks we provided above
+static void replaceChunkHooks(void) {
+  unsigned narenas;
+  unsigned arena;
+
+  // set the pointers for the new_hooks to our above functions
+  chunk_hooks_t new_hooks = {
+    chunk_alloc,
+    null_dalloc,
+    null_commit,
+    null_decommit,
+    null_purge,
+    null_split,
+    null_merge
+  };
+
+  // for each arena, change the chunk hooks
+  narenas = get_num_arenas();
+  for (arena=0; arena<narenas; arena++) {
+    char path[128];
+    snprintf(path, 128, "arena.%u.chunk_hooks", arena);
```

Comment:
Nit-pick, nit-pick, but when I use snprintf() like this I like to say
    snprintf(path, sizeof(path), ...);
instead of 128 (for example), because then someone can change the size of the target buffer decl and the snprintf will follow right along. I.e., it avoids the need to maintain the duplicated value '128'.